Insert failed on MySQL VARBINARY column when it is sharding column

## Feature Request

### Is your feature request related to a problem?
Yes

### Describe the feature you would like.
MySQL is used, table primary key is VARBINARY type. Primary key column is sharding column, and keyGenerateStrategy is NOT configured.
When insert into table, there's exception thrown.
It's tested in migration job. The inserted varbinary value is byte array (read from MySQL and write to ShardingSphere with MySQL).
```
o.a.s.e.e.h.g.LogJobErrorHandler - Job 'j0101b36ade21a62d0d8bdc8b55d2eb1ea55a' exception occur in job processing
java.lang.RuntimeException: java.util.concurrent.ExecutionException: org.apache.shardingsphere.sharding.exception.data.NotImplementComparableValueException: Sharding value: [B@6fd51628 value must implements Comparable.
        at org.apache.shardingsphere.data.pipeline.core.execute.ExecuteEngine.trigger(ExecuteEngine.java:117)
        at org.apache.shardingsphere.data.pipeline.core.task.InventoryIncrementalTasksRunner.executeInventoryTask(InventoryIncrementalTasksRunner.java:97)
        at org.apache.shardingsphere.data.pipeline.core.task.InventoryIncrementalTasksRunner.start(InventoryIncrementalTasksRunner.java:84)
        at org.apache.shardingsphere.data.pipeline.core.job.AbstractSimplePipelineJob.execute(AbstractSimplePipelineJob.java:60)
        at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:33)
        at org.apache.shardingsphere.elasticjob.simple.executor.SimpleJobExecutor.process(SimpleJobExecutor.java:29)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:173)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.process(ElasticJobExecutor.java:142)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:124)
        at org.apache.shardingsphere.elasticjob.executor.ElasticJobExecutor.execute(ElasticJobExecutor.java:100)
        at org.apache.shardingsphere.elasticjob.lite.internal.schedule.LiteJob.execute(LiteJob.java:35)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: java.util.concurrent.ExecutionException: org.apache.shardingsphere.sharding.exception.data.NotImplementComparableValueException: Sharding value: [B@6fd51628 value must implements Comparable.
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
        at org.apache.shardingsphere.data.pipeline.core.execute.ExecuteEngine.trigger(ExecuteEngine.java:111)
        ... 12 common frames omitted
Caused by: org.apache.shardingsphere.sharding.exception.data.NotImplementComparableValueException: Sharding value: [B@6fd51628 value must implements Comparable.
        at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.InsertClauseShardingConditionEngine.lambda$getShardingValue$0(InsertClauseShardingConditionEngine.java:145)
        at org.apache.shardingsphere.infra.util.exception.ShardingSpherePreconditions.checkState(ShardingSpherePreconditions.java:41)
        at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.InsertClauseShardingConditionEngine.getShardingValue(InsertClauseShardingConditionEngine.java:145)
        at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.InsertClauseShardingConditionEngine.createShardingCondition(InsertClauseShardingConditionEngine.java:115)
        at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.InsertClauseShardingConditionEngine.createShardingConditionsWithInsertValues(InsertClauseShardingConditionEngine.java:84)
        at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.InsertClauseShardingConditionEngine.createShardingConditions(InsertClauseShardingConditionEngine.java:71)
        at org.apache.shardingsphere.sharding.route.engine.condition.engine.impl.DefaultShardingConditionEngine.createShardingConditions(DefaultShardingConditionEngine.java:47)
        at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createShardingConditions(ShardingSQLRouter.java:69)
        at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:53)
        at org.apache.shardingsphere.sharding.route.engine.ShardingSQLRouter.createRouteContext(ShardingSQLRouter.java:46)
        at org.apache.shardingsphere.infra.route.engine.impl.PartialSQLRouteExecutor.route(PartialSQLRouteExecutor.java:69)
        at org.apache.shardingsphere.infra.route.engine.SQLRouteEngine.route(SQLRouteEngine.java:55)
        at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.route(KernelProcessor.java:59)
        at org.apache.shardingsphere.infra.context.kernel.KernelProcessor.generateExecutionContext(KernelProcessor.java:51)
        at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.createExecutionContext(ShardingSpherePreparedStatement.java:697)
        at org.apache.shardingsphere.driver.jdbc.core.statement.ShardingSpherePreparedStatement.addBatch(ShardingSpherePreparedStatement.java:780)
        at org.apache.shardingsphere.data.pipeline.core.importer.DataSourceImporter.executeBatchInsert(DataSourceImporter.java:191)
        at org.apache.shardingsphere.data.pipeline.core.importer.DataSourceImporter.doFlush(DataSourceImporter.java:160)
        at org.apache.shardingsphere.data.pipeline.core.importer.DataSourceImporter.tryFlush(DataSourceImporter.java:142)
        at org.apache.shardingsphere.data.pipeline.core.importer.DataSourceImporter.flushInternal(DataSourceImporter.java:135)
        at org.apache.shardingsphere.data.pipeline.core.importer.DataSourceImporter.flush(DataSourceImporter.java:125)
        at org.apache.shardingsphere.data.pipeline.core.importer.DataSourceImporter.runBlocking(DataSourceImporter.java:101)
        at org.apache.shardingsphere.data.pipeline.api.executor.AbstractLifecycleExecutor.start(AbstractLifecycleExecutor.java:52)
        at org.apache.shardingsphere.data.pipeline.api.executor.AbstractLifecycleExecutor.run(AbstractLifecycleExecutor.java:90)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run$$$capture(CompletableFuture.java:1736)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

Example table structure, yaml configuration and test code could refer to #24135.

Just remove keyGenerateStrategy and change `shardingColumn: user_id` to `shardingColumn: order_id`.
```
databaseName: logic_db
# ...

rules:
  - !SHARDING
    autoTables:
      t_order:
        actualDataSources: ds_2,ds_3,ds_4
        logicTable: t_order
        shardingStrategy:
          standard:
            shardingAlgorithmName: t_order_hash_mod
            shardingColumn: order_id
    shardingAlgorithms:
      t_order_hash_mod:
        props:
          sharding-count: '6'
        type: hash_mod
```

--- a/features/sharding/api/src/main/java/org/apache/shardingsphere/sharding/api/sharding/standard/PreciseShardingValue.java
+++ b/features/sharding/api/src/main/java/org/apache/shardingsphere/sharding/api/sharding/standard/PreciseShardingValue.java
@@ -20,8 +20,8 @@ package org.apache.shardingsphere.sharding.api.sharding.standard;
 import lombok.Getter;
 import lombok.RequiredArgsConstructor;
 import lombok.ToString;
-import org.apache.shardingsphere.sharding.api.sharding.ShardingValue;
 import org.apache.shardingsphere.infra.datanode.DataNodeInfo;
+import org.apache.shardingsphere.sharding.api.sharding.ShardingValue;
 
 /**
  * Sharding value for precise.
@@ -29,7 +29,7 @@ import org.apache.shardingsphere.infra.datanode.DataNodeInfo;
 @RequiredArgsConstructor
 @Getter
 @ToString
-public final class PreciseShardingValue<T extends Comparable<?>> implements ShardingValue {
+public final class PreciseShardingValue<T> implements ShardingValue {
     
     private final String logicTableName;
     

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/mod/HashModShardingAlgorithm.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/mod/HashModShardingAlgorithm.java
@@ -58,7 +58,7 @@ public final class HashModShardingAlgorithm implements StandardShardingAlgorithm
         return availableTargetNames;
     }
     
-    private long hashShardingValue(final Comparable<?> shardingValue) {
+    private long hashShardingValue(final Object shardingValue) {
         return Math.abs((long) shardingValue.hashCode());
     }
     

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/engine/condition/engine/InsertClauseShardingConditionEngine.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/engine/condition/engine/InsertClauseShardingConditionEngine.java
@@ -24,8 +24,6 @@ import org.apache.shardingsphere.infra.binder.segment.insert.values.InsertValueC
 import org.apache.shardingsphere.infra.binder.statement.dml.InsertStatementContext;
 import org.apache.shardingsphere.infra.binder.statement.dml.SelectStatementContext;
 import org.apache.shardingsphere.infra.metadata.database.ShardingSphereDatabase;
-import org.apache.shardingsphere.infra.util.exception.ShardingSpherePreconditions;
-import org.apache.shardingsphere.sharding.exception.data.NotImplementComparableValueException;
 import org.apache.shardingsphere.sharding.exception.data.NullShardingValueException;
 import org.apache.shardingsphere.sharding.route.engine.condition.ExpressionConditionUtils;
 import org.apache.shardingsphere.sharding.route.engine.condition.ShardingCondition;
@@ -134,13 +132,10 @@ public final class InsertClauseShardingConditionEngine {
         }
     }
     
-    @SuppressWarnings("rawtypes")
-    private Comparable<?> getShardingValue(final SimpleExpressionSegment expressionSegment, final List<Object> params) {
-        Object result = expressionSegment instanceof ParameterMarkerExpressionSegment
+    private Object getShardingValue(final SimpleExpressionSegment expressionSegment, final List<Object> params) {
+        return expressionSegment instanceof ParameterMarkerExpressionSegment
                 ? params.get(((ParameterMarkerExpressionSegment) expressionSegment).getParameterMarkerIndex())
                 : ((LiteralExpressionSegment) expressionSegment).getLiterals();
-        ShardingSpherePreconditions.checkState(result instanceof Comparable, () -> new NotImplementComparableValueException("Sharding", result));
-        return (Comparable) result;
     }
     
     private List<ShardingCondition> createShardingConditionsWithInsertSelect(final InsertStatementContext sqlStatementContext, final List<Object> params) {

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/engine/condition/value/ListShardingConditionValue.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/engine/condition/value/ListShardingConditionValue.java
@@ -31,7 +31,7 @@ import java.util.stream.Collectors;
  */
 @RequiredArgsConstructor
 @Getter
-public final class ListShardingConditionValue<T extends Comparable<?>> implements ShardingConditionValue {
+public final class ListShardingConditionValue<T> implements ShardingConditionValue {
     
     private final String columnName;
     

--- a/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/strategy/type/standard/StandardShardingStrategy.java
+++ b/features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/route/strategy/type/standard/StandardShardingStrategy.java
@@ -71,7 +71,7 @@ public final class StandardShardingStrategy implements ShardingStrategy {
     @SuppressWarnings({"unchecked", "rawtypes"})
     private Collection<String> doSharding(final Collection<String> availableTargetNames, final ListShardingConditionValue<?> shardingValue, final DataNodeInfo dataNodeInfo) {
         Collection<String> result = new LinkedList<>();
-        for (Comparable<?> each : shardingValue.getValues()) {
+        for (Object each : shardingValue.getValues()) {
             String target = shardingAlgorithm.doSharding(availableTargetNames,
                     new PreciseShardingValue(shardingValue.getTableName(), shardingValue.getColumnName(), dataNodeInfo, each));
             if (null != target && availableTargetNames.contains(target)) {

--- a/features/sharding/plugin/cache/src/main/java/org/apache/shardingsphere/sharding/cache/checker/ShardingRouteCacheableChecker.java
+++ b/features/sharding/plugin/cache/src/main/java/org/apache/shardingsphere/sharding/cache/checker/ShardingRouteCacheableChecker.java
@@ -186,7 +186,7 @@ public final class ShardingRouteCacheableChecker {
     
     private static boolean isConditionTypeCacheable(final ShardingConditionValue conditionValue) {
         if (conditionValue instanceof ListShardingConditionValue<?>) {
-            for (Comparable<?> eachValue : ((ListShardingConditionValue<?>) conditionValue).getValues()) {
+            for (Object eachValue : ((ListShardingConditionValue<?>) conditionValue).getValues()) {
                 if (!(eachValue instanceof Number)) {
                     return false;
                 }

--- a/jdbc/core/src/test/java/org/apache/shardingsphere/driver/ShardingSphereDriverTest.java
+++ b/jdbc/core/src/test/java/org/apache/shardingsphere/driver/ShardingSphereDriverTest.java
@@ -22,6 +22,7 @@ import org.junit.Test;
 
 import java.sql.Connection;
 import java.sql.DriverManager;
+import java.sql.PreparedStatement;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Statement;
@@ -50,6 +51,26 @@ public final class ShardingSphereDriverTest {
             try (ResultSet resultSet = statement.executeQuery("SELECT COUNT(1) FROM t_order")) {
                 assertTrue(resultSet.next());
                 assertThat(resultSet.getInt(1), is(2));
+            }
+        }
+    }
+    
+    @Test
+    public void assertVarbinaryColumnWorks() throws SQLException {
+        try (
+                Connection connection = DriverManager.getConnection("jdbc:shardingsphere:classpath:config/driver/foo-driver-fixture.yaml");
+                Statement statement = connection.createStatement()) {
+            assertThat(connection, instanceOf(ShardingSphereConnection.class));
+            statement.execute("DROP TABLE IF EXISTS t_order");
+            statement.execute("CREATE TABLE t_order (order_id VARBINARY(64) PRIMARY KEY, user_id INT)");
+            PreparedStatement preparedStatement = connection.prepareStatement("INSERT INTO t_order (order_id, user_id) VALUES (?, ?)");
+            preparedStatement.setBytes(1, new byte[]{-1, 0, 1});
+            preparedStatement.setInt(2, 101);
+            int updatedCount = preparedStatement.executeUpdate();
+            assertThat(updatedCount, is(1));
+            try (ResultSet resultSet = statement.executeQuery("SELECT COUNT(1) FROM t_order")) {
+                assertTrue(resultSet.next());
+                assertThat(resultSet.getInt(1), is(1));
             }
         }
     }

--- a/jdbc/core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/driver/ShardingSphereDriverURLManagerTest.java
+++ b/jdbc/core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/driver/ShardingSphereDriverURLManagerTest.java
@@ -27,7 +27,7 @@ import static org.hamcrest.MatcherAssert.assertThat;
 
 public final class ShardingSphereDriverURLManagerTest {
     
-    private final int fooDriverConfigLength = 995;
+    private final int fooDriverConfigLength = 999;
     
     @Test(expected = DriverURLProviderNotFoundException.class)
     public void assertNewConstructorWithEmptyURL() {

--- a/jdbc/core/src/test/resources/config/driver/foo-driver-fixture.yaml
+++ b/jdbc/core/src/test/resources/config/driver/foo-driver-fixture.yaml
@@ -41,11 +41,11 @@ rules:
             shardingColumn: order_id
             shardingAlgorithmName: auto-mod
         keyGenerateStrategy:
-          column: order_id
+          column: user_id
           keyGeneratorName: snowflake
     shardingAlgorithms:
       auto-mod:
-        type: MOD
+        type: HASH_MOD
         props:
           sharding-count: 2
     

--- a/test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/migration/primarykey/IndexesMigrationE2EIT.java
+++ b/test/e2e/pipeline/src/test/java/org/apache/shardingsphere/test/e2e/data/pipeline/cases/migration/primarykey/IndexesMigrationE2EIT.java
@@ -52,6 +52,12 @@ import static org.hamcrest.Matchers.is;
 @Slf4j
 public final class IndexesMigrationE2EIT extends AbstractMigrationE2EIT {
     
+    private static final String ORDER_TABLE_SHARDING_RULE_FORMAT = "CREATE SHARDING TABLE RULE t_order(\n"
+            + "STORAGE_UNITS(ds_2,ds_3,ds_4),\n"
+            + "SHARDING_COLUMN=%s,\n"
+            + "TYPE(NAME=\"hash_mod\",PROPERTIES(\"sharding-count\"=\"6\"))\n"
+            + ");";
+    
     public IndexesMigrationE2EIT(final PipelineTestParameter testParam) {
         super(testParam);
     }
@@ -86,7 +92,7 @@ public final class IndexesMigrationE2EIT extends AbstractMigrationE2EIT {
         } else {
             return;
         }
-        assertMigrationSuccess(sql, new UUIDKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
+        assertMigrationSuccess(sql, "user_id", new UUIDKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
     }
     
     @Test
@@ -99,7 +105,7 @@ public final class IndexesMigrationE2EIT extends AbstractMigrationE2EIT {
         } else {
             return;
         }
-        assertMigrationSuccess(sql, new UUIDKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
+        assertMigrationSuccess(sql, "user_id", new UUIDKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
     }
     
     @Test
@@ -112,7 +118,7 @@ public final class IndexesMigrationE2EIT extends AbstractMigrationE2EIT {
         } else {
             return;
         }
-        assertMigrationSuccess(sql, new SnowflakeKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
+        assertMigrationSuccess(sql, "user_id", new SnowflakeKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
     }
     
     @Test
@@ -126,10 +132,11 @@ public final class IndexesMigrationE2EIT extends AbstractMigrationE2EIT {
         } else {
             return;
         }
-        assertMigrationSuccess(sql, new UUIDKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
+        assertMigrationSuccess(sql, "order_id", new UUIDKeyGenerateAlgorithm(), consistencyCheckAlgorithmType);
     }
     
-    private void assertMigrationSuccess(final String sqlPattern, final KeyGenerateAlgorithm generateAlgorithm, final String consistencyCheckAlgorithmType) throws SQLException, InterruptedException {
+    private void assertMigrationSuccess(final String sqlPattern, final String shardingColumn, final KeyGenerateAlgorithm generateAlgorithm,
+                                        final String consistencyCheckAlgorithmType) throws SQLException, InterruptedException {
         initEnvironment(getDatabaseType(), new MigrationJobType());
         sourceExecuteWithLog(String.format(sqlPattern, getSourceTableOrderName()));
         try (Connection connection = getSourceDataSource().getConnection()) {
@@ -138,7 +145,7 @@ public final class IndexesMigrationE2EIT extends AbstractMigrationE2EIT {
         addMigrationProcessConfig();
         addMigrationSourceResource();
         addMigrationTargetResource();
-        createTargetOrderTableRule();
+        proxyExecuteWithLog(String.format(ORDER_TABLE_SHARDING_RULE_FORMAT, shardingColumn), 2);
         startMigration(getSourceTableOrderName(), getTargetTableOrderName());
         String jobId = listJobId().get(0);
         waitJobPrepareSuccess(String.format("SHOW MIGRATION STATUS '%s'", jobId));

